internal/service/alarm_subscription_handler.go

```diff
@@ -358,3 +402,206 @@ func (h *alarmSubscriptionHandler) mapItem(ctx context.Context,
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	value, ok := configmap.Data[entry_key]
+	if !ok {
+		err = ErrNotFound
+	}
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) ConfigMapDeleteEntry(
+	ctx context.Context, entry_key string) (err error) {
+
+	//test to read the configmap
+	configmap := &corev1.ConfigMap{}
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	_, ok := configmap.Data[entry_key]
+
+	//entry not found
+	if !ok {
+		return
+	}
+	delete(configmap.Data, entry_key)
+
+	// Note: there is a case configmap is empty shall we remove the configmap
+	err = h.hubClient.Update(ctx, configmap, &clnt.UpdateOptions{})
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) RecoveryFromConfigMap(ctx context.Context) (err error) {
+
+	//test to read the configmap
+	configmap := &corev1.ConfigMap{}
+
+	key := clnt.ObjectKey{
+		Namespace: TESTNAMESPACE,
+		Name:      TESTCONFIGMAPNAME,
+	}
+	err = h.hubClient.Get(ctx, key, configmap)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		//there is error and err is not notfound error
+		//panic("unexpected error")
+		return
+	}
+
+	if configmap.Data == nil {
+		return
+	}
+
+	for mapkey, value := range configmap.Data {
+		var object data.Object
+		err = jsoniter.Unmarshal([]byte(value), &object)
+		if err != nil {
+			continue
+		}
+
+		h.subscriptionMap[mapkey] = object
+
+	}
+
+	return
+}
+
+func (h *alarmSubscriptionHandler) WatchConfigMap(ctx context.Context) (err error) {
+
+	raw_opt := metav1.SingleObject(metav1.ObjectMeta{Name: TESTCONFIGMAPNAME, Namespace: TESTNAMESPACE})
+	opt := clnt.ListOptions{}
+	opt.Raw = &raw_opt
+
+	watcher, err := h.hubClient.Watch(ctx, &corev1.ConfigMapList{}, &opt)
+
+	if err != nil {
+		panic("failed to launch watcher")
+	}
+
+	go func() {
+		for {
+			event, open := <-watcher.ResultChan()
+			if open {
+				switch event.Type {
+				case watch.Added:
+					fallthrough
+				case watch.Modified:
+					h.subscritionMapMemoryLock.Lock()
+					configmap, ok := event.Object.(*corev1.ConfigMap)
+					if ok {
+						h.subscriptionMap = make(map[string]data.Object)
+						for k, value := range configmap.Data {
+							var object data.Object
+							err = jsoniter.Unmarshal([]byte(value), &object)
+							if err != nil {
+								continue
+							}
+							h.subscriptionMap[k] = object
+						}
+					}
+					h.subscritionMapMemoryLock.Unlock()
+				case watch.Deleted:
+					h.subscritionMapMemoryLock.Lock()
```

Comment:
Same here, use the `defer` mechanism even if it is not strictly needed. In this case this code can't possible fail (unless if memory is exhausted) but it may be changed in the future and cause a locking bug.

Reply:
As this is infinite loop, handle the error path differently.